src/main/java/li/cil/sedna/riscv/gdbstub/GDBStub.java

```diff
@@ -325,17 +362,121 @@ private void handleBreakpointAdd(final ByteBuffer buffer) throws IOException {
+        final int length = Integer.parseInt(command, addressCharsEnd + 1, command.length(), 16);
+        final Interval interval = new Interval(address, length);
+        try (final var s = new GDBPacketOutputStream(output);
+             final var w = new OutputStreamWriter(s, StandardCharsets.US_ASCII)) {
+            Watchpoint watchpoint =  switch (type) {
+                case '2' -> new Watchpoint(interval, false, true);
+                case '3' -> new Watchpoint(interval, true, false);
+                case '4' -> new Watchpoint(interval, true, true);
+                default -> throw new IllegalStateException("Unexpected value: " + type);
+            };
+            cpu.removeWatchpoint(watchpoint);
+            w.write("OK");
+        }
+    }
+
+    private MessageStop handleStep() {
         cpu.step();
         state = GDBState.STOP_REPLY;
+        return new MessageStop();
+    }
+
+    private void handleQuery(ByteBuffer packetBuffer) throws IOException {
+        final byte[] Supported = "Supported:".getBytes(StandardCharsets.US_ASCII);
+        final byte[] Attached = "Attached".getBytes(StandardCharsets.US_ASCII);
+        final byte[] features = "Xfer:features:read:".getBytes(StandardCharsets.US_ASCII);
+        if (ByteBufferUtils.startsWith(packetBuffer, ByteBuffer.wrap(Supported))) {
+            packetBuffer.position(packetBuffer.position() + Supported.length);
+            handleSupported(packetBuffer);
+        } else if (ByteBufferUtils.startsWith(packetBuffer, ByteBuffer.wrap(Attached))) {
+            try (final var s = new GDBPacketOutputStream(output);
+                 final var w = new OutputStreamWriter(s, StandardCharsets.US_ASCII)) {
+                w.write("1");
+            }
+        } else if (ByteBufferUtils.startsWith(packetBuffer, ByteBuffer.wrap(features))) {
+            packetBuffer.position(packetBuffer.position() + features.length);
+            handleReadTargetDescription(packetBuffer);
+        } else {
+            unknownCommand(packetBuffer);
+        }
+    }
+
+    private void handleReadTargetDescription(ByteBuffer buf) throws IOException {
+        try (final var s = new GDBPacketOutputStream(output)) {
+            try {
+                String annex = ByteBufferUtils.getStringToken(buf, (byte) ':');
+                int offset = Integer.parseInt(ByteBufferUtils.getStringToken(buf, (byte) ','), 16);
+                int length = Integer.parseInt(ByteBufferUtils.tokenAsString(buf), 16);
+                handleReadTargetDescription(annex, offset, length, s);
+            } catch (ByteBufferUtils.TokenException e) {
+                LOGGER.error("Failed to parse qXfer features read packet", e);
+                s.write("E00".getBytes(StandardCharsets.US_ASCII));
+            }
+        }
+    }
+
+    private void handleReadTargetDescription(String annex, int offset, int length, OutputStream out) throws IOException {
+        if(!annex.equals("target-riscv64.xml")) {
```

Comment:
I think this should be `target.xml` - that's what my `riscv64-linux-gnu-gdb` requests, and that's what docs mention: "The main description is always loaded from the ‘target.xml’ annex." (https://sourceware.org/gdb/current/onlinedocs/gdb.html/General-Query-Packets.html under "qXfer:features:read:annex")

With this change I'm able to read floating point registers.

Reply:
Oops yes I broke that in b1c0dad5091e0d6ed6026307f21ed4bbdfc5d293
Feel free to add a "suggestion" with the fix and I'll approve it. Or I'll eventually get around to it